Setup tests and cookie

JSAssassin opened the pull request:
Addresses #6 

Edit: Need to update bedrock-authn-token and bedrock-account once these PRs are merged.
https://github.com/digitalbazaar/bedrock-account/pull/25
https://github.com/digitalbazaar/bedrock-authn-token/pull/33

JSAssassin committed:
Set up tests.

JSAssassin committed:
Add some tests.

JSAssassin committed:
Set a cookie and bake clientID into the cookie.

- Add additional tests.

mattcollier commented on lib/config.js:
```diff
@@ -8,6 +8,13 @@ require('bedrock-validation');
 
 const cfg = config['authn-token-http'] = {};
 
+cfg.cid = {
```
needs some documentation

JSAssassin replied:
Updated.

mattcollier commented on lib/index.js:
```diff
@@ -54,17 +55,17 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
         res.status(204).end();
       } else {
-        // all other types require authentication to create
+        // all other types require authentication to create(Require an account)
```
the additional parenthetical here make things less clear to me rather than more clear.  Please reword to make the meaning more clear.  Although I'm not sure specifically calling out that an "account" is required, something like this might be better:
```suggestion
        // all other types require an authenticated account to create
```

JSAssassin replied:
Updated.

mattcollier commented on lib/index.js:
```diff
@@ -92,14 +94,12 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
     asyncHandler(async (req, res) => {
       const {type} = req.params;
       const {account, email} = req.query;
+      // if token is undefined `brAuthnToken.get` throws a 404 error
```
```suggestion
      // if a token is does not exist `brAuthnToken.get` throws a NotFoundError
```

JSAssassin replied:
Updated.

mattcollier commented on lib/index.js:
```diff
@@ -128,15 +128,19 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
-          authenticatedMethods.push('token-client-registration');
+        // get clientId from cookie
+        const clientId = req.cookies.cid;
+        // if clientId exists, check if client has been registered. If it is
+        // registered, then add `token-client-registration` to the list of
+        // `authenticatedMethods`.
```
I think remove this, don't think this comment adds any value in addition to the simple code.

JSAssassin replied:
Updated.

mattcollier commented on lib/index.js:
```diff
@@ -157,10 +161,16 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
+        // set a cookie in the response
+        const maxAge = cfg.cid.defaults.maxAge;
+        res.setHeader('Set-Cookie', `cid=${clientId}; expires=${maxAge};` +
+          `HttpOnly`);
```
No need for backticks here.  Infact style convention prohibits backticks without template string.

JSAssassin replied:
Updated.

mattcollier commented on lib/index.js:
```diff
@@ -256,10 +266,15 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
+        isRegistered = result.registered;
+      }
+      res.json({registered: isRegistered});
+
```
remove extra whitespace.

JSAssassin replied:
Updated.

mattcollier commented on test/mocha/.eslintrc:
```diff
@@ -4,6 +4,7 @@
     "assertNoError": true,
-    "should": true
+    "should": true,
+    "passportStub": true
```
I don't think `passportStub` is a global

JSAssassin replied:
This is the example I was following https://github.com/digitalbazaar/bedrock-account-http/blob/master/test/mocha/.eslintrc#L8.  I tried removing the `passportStub` from here and moving https://github.com/digitalbazaar/bedrock-authn-token-http/pull/13/files#diff-c1129c8b045390789fa8ff62f2c6b4a9R5-R7 into `10-api.js` file, but some of the tests fail when I do that.

mattcollier replied:
Here's another example: https://github.com/digitalbazaar/bedrock-profile-http/blob/master/test/mocha/10-api.js#L22 that does not involve a global.

JSAssassin replied:
@mattcollier On further investigation, it looks like the reason why the stub(for optionallyAuthenticated function) was failing for some of the tests when it is not global and passes when stub is set to global is that some api endpoints has this `optionallyAuthenticated` check  https://github.com/digitalbazaar/bedrock-authn-token-http/blob/master/lib/index.js#L48, while others have `ensureAuthenticated` check
https://github.com/digitalbazaar/bedrock-authn-token-http/blob/master/lib/index.js#L109

The stub works for `ensureAuthenticated` which is an api from brPassport that wraps/calls `optionallyAuthenticated` api https://github.com/digitalbazaar/bedrock-passport/blob/master/lib/index.js#L294 even when it is not set to global.

And doesn’t work for `optionallyAuthenticated` https://github.com/digitalbazaar/bedrock-passport/blob/master/lib/index.js#L273 This seems to be because express keeps a reference to the original `optionallyAuthenticated` function before sinon has a chance to stub it. I think it works for ensureAuthenticated because we don’t stub that directly, but stub optionallyAuthenticated instead. So express keeps a reference to ensureAuthenticated which calls the stub.

The solution to the problem would be either to  set the stub before starting express or make another api in brPassport that would call `optionallyAuthenticated` 

```
	api.foo = async (req, res, next) => {
  	    api.optionallyAuthenticated(req, res, next);
	};
```

And replace `optionallyAuthenticated` check with `foo`  in bedrock-auth-token-http

mattcollier replied:
Thanks for doing all this research/investigation.  If the stub is setup in test.js it does or does not work?

JSAssassin replied:
@mattcollier yes, it will work.  I think if we do it before starting bedrock.

mattcollier replied:
OK, please do it that way (with the global again if necessary) with a detailed comment that captures all the work you did to explain to future visitors why it's being done in this way.

aljones15 replied:
just so you know the stubs on `optionallyAuthenticated` and `ensureAuthenticated` can fail. So if any further issues occurs try stubbing `api.authenticate` in `bedrock-passport`: https://github.com/digitalbazaar/bedrock-passport/blob/33759c310195712f7ce763e063ffe9f7a8c92d9d/lib/index.js#L52-L72

That is the method that sets user for both `optionallyAuthenticated` and `ensureAuthenticated` so it should work for both methods. All you need to do is make sure it returns the correct `{user}`

EDIT: one small issue here though is that stubbing `authenticate` might not stub `req.actor` or lead to an issue with the actor.

mattcollier commented on test/mocha/10-api.js:
```diff
@@ -1,3 +1,505 @@
+`${config['authn-token-http'].routes.registration}`;
+
+// use agent to avoid self-signed certificate errors
+const agent = new https.Agent({rejectUnauthorized: false});
```
this should use `bedrock-http-agent` and corresponding entries in test.config.js which I see are already included in the test config here.

JSAssassin replied:
I followed this as a reference. https://github.com/digitalbazaar/bedrock-profile-http/blob/ed90c26c6ce17b593bb08f7f40070c74a6e95524/test/mocha/10-api.js#L31

Is there an example on how to do this differently?

mattcollier replied:
Here you go: https://github.com/digitalbazaar/bedrock-edv-storage/blob/abc79792fb7eda7725e582a5c9dd1bf931ce4d0a/test/mocha/16-update.js#L76-L77

mattcollier commented on test/mocha/10-api.js:
```diff
@@ -1,3 +1,505 @@
+      const type = 'password';
+      const accountId = accounts['alpha@example.com'].account.id;
+      const hash =
+      '$2y$12$LR16D./bm4rN7a5PHhRypeNs5SW4KNWz.XjSvw7JG6KYd1yKjcGqm';
```
indent.

mattcollier reviewed (changes requested):
Looking good.

JSAssassin committed:
Fix comments, style issues and add additional test.

JSAssassin committed:
Get `agent` from `bedrock-https-agent`.

dlongley commented on lib/config.js:
```diff
@@ -8,6 +8,14 @@ require('bedrock-validation');
+    // 30 days expiration
+    maxAge: 30 * 24 * 60 * 60 * 1000,
+  }
+};
```
There might be a need to configure the name of this cookie, so we shouldn't make its name the same thing as the configuration variable. Also, we need a short name for the cookie itself to save overhead on network data, but we don't need that here in the config, so we should be verbose. We might also want some future proofing here, so let's name space the client cookie info under `cookies` as well. Lastly, do we need "defaults"? I've removed that field in my suggestion.

```suggestion
cfg.cookies = {};
cfg.cookies.clientId = {
  name: 'cid',
  // 30 days expiration
  maxAge: 30 * 24 * 60 * 60 * 1000
};
```

mattcollier commented on test/mocha/10-api.js:
```diff
@@ -5,13 +5,14 @@
 const brAuthnToken = require('bedrock-authn-token');
 const {authenticator} = require('otplib');
+const brHttpsAgent = require('bedrock-https-agent');
+const {agent} = brHttpsAgent;
```
I want to be sure that you understand that extracting the `agent` in the header like this only works in a test spec because of how bedrock bootstraps itself and order of operation.  It is not safe to do this in normal library code.  Because we don't want people thinking it's OK to do this everywhere, I think it's best to extract the `agent` immediately prior to its use in the tests below.

JSAssassin replied:
Updated.

JSAssassin committed:
Make cookie name configurable and use express `res.cookie` to set it.

- Fix comment styles.
- Extract agent prior to its use in the tests.
- Add `options` to `cfg.cookies.clientId`.

dlongley commented on lib/index.js:
```diff
@@ -54,17 +55,17 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
-
       if(type === 'nonce') {
-        // permitted without authentication
+        // permitted without authentication. Do not require an account since
+        // anyone can request for a nonce.
```
Our comment style is that if we need full punctuation, capitalization, and complete sentences/paragraphs, that sort of thing -- to describe what's going on, we should do a full multiline comment (`/* Some sentences and a length explanation goes here. */`). If not, we do the simpler inline comments that don't need all that. Since this style mixed the two together, I made the suggestion below. I also tweaked the language a little bit to remove the implication 

```suggestion
        // since anyone can request a nonce, this is permitted without authentication
```

dlongley commented on lib/index.js:
```diff
@@ -54,17 +55,17 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
         res.status(204).end();
       } else {
-        // all other types require authentication to create
+        // all other types require an authenticated account to create.
```
I wouldn't change this, because `authentication` doesn't necessarily mean you have an account, it just means you are a known user of the system. There are bedrock systems where a user might actually not have "an account" but they are dynamically generated a `user`/`actor` on the system after authenticating.

```suggestion
        // all other types require authentication to create
```

mattcollier replied:
I think incorporating some of what longley has just said into the comment will prevent confusion in the future.

dlongley commented on lib/index.js:
```diff
@@ -128,15 +128,16 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
-        if(result.registered) {
-          authenticatedMethods.push('token-client-registration');
+        // get clientId from cookie
+        const clientId = req.cookies.cid;
```
Can use the cookie name from the config here.

dlongley commented on lib/index.js:
```diff
@@ -157,10 +158,16 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
         await brAuthnToken.clients.set(
           {actor: null, email, account, clientId, authenticated: true});
+        // set a cookie in the response
+        const maxAge = cfg.cid.defaults.maxAge;
```
Config update per the above suggestion.

dlongley commented on lib/index.js:
```diff
@@ -157,10 +158,16 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
           {actor: null, email, account, clientId, authenticated: true});
+        // set a cookie in the response
+        const maxAge = cfg.cid.defaults.maxAge;
+        res.setHeader('Set-Cookie', `cid=${clientId}; expires=${maxAge};` +
+          'HttpOnly');
```
Is `expires` the right param here for `maxAge`? Also, does `res` have an API for setting cookies or do we have to do it via the headers API?

JSAssassin replied:
I have updated it to use https://expressjs.com/en/5x/api.html#res.cookie. You are right `expires` is not the right param for maxAge. The value for `expires` should be a date, I removed it . There was a `maxAge` param for maxAge, so used that.

dlongley reviewed (changes requested):
Coming along nicely!

dlongley reviewed (approved):
LGTM! It does look like we don't have any `/login` tests, however. Are those going to be added in another PR or should they be added here? Tests should be exercising both pass/fail for authenticating with multiple factors and then logging in. This would include exercising the new client ID code as part of that process.

dlongley commented on lib/index.js:
```diff
@@ -157,10 +159,14 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
           {actor: null, email, account, clientId, authenticated: true});
+        // set a cookie in the response
+        const {name, maxAge} = cfg.cookies.clientId;
+        res.cookie(name, clientId, {maxAge, httpOnly: true});
```
Actually, we should change this so it's easier to configure this cookie in other systems. We should just be passing the options in directly from the config. I'll make a note up where we set the config about this as well.

dlongley commented on lib/config.js:
```diff
@@ -8,6 +8,14 @@ require('bedrock-validation');
+// client id to be baked into cookies that is used in client registration
+cfg.cookies.clientId = {
+  name: 'cid',
+  // 30 days expiration
+  maxAge: 30 * 24 * 60 * 60 * 1000
```
We should nest the cookie options under `options` so that they can be configured more easily by applications. Some applications might need this cookie to be `SameSite=None` for example.

```suggestion
  options: {
    // 30 days expiration
    maxAge: 30 * 24 * 60 * 60 * 1000,
    httpOnly: true
  }
```

JSAssassin committed:
Add tests for login route.

- Add comments and set passport stub before bedrock starts.

dlongley commented on lib/config.js:
```diff
@@ -8,6 +8,16 @@ require('bedrock-validation');
+    name: 'cid',
+    // 30 days expiration
+    maxAge: 30 * 24 * 60 * 60 * 1000
+  }
```
Sorry, I was unclear about what to do here. We want the name to be separate from the options ... so we can just pass the options directly into the `res.cookie` API below.

```suggestion
  name: 'cid',
  options: {
    // 30 days expiration
    maxAge: 30 * 24 * 60 * 60 * 1000,
    httpOnly: true
  }
```

dlongley commented on lib/index.js:
```diff
@@ -157,10 +159,14 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
           {actor: null, email, account, clientId, authenticated: true});
+        // set a cookie in the response
+        const {name, maxAge} = cfg.cookies.clientId.options;
+        res.cookie(name, clientId, {maxAge, httpOnly: true});
```
```suggestion
        const {name, options} = cfg.cookies.clientId;
        res.cookie(name, clientId, options);
```

JSAssassin committed:
Make `name` to be separate from `options` in config.

closes #6

dlongley reviewed (approved):
Thanks @JSAssassin!

aljones15 commented on test/mocha/10-api.js:
```diff
@@ -1,3 +1,745 @@
+const {httpClient} = require('@digitalbazaar/http-client');
+const brAccount = require('bedrock-account');
+const brAuthnToken = require('bedrock-authn-token');
+const {authenticator} = require('otplib');
```
This is not required in this PR, but we might consider the async version of authenticator in the future:

https://github.com/yeojz/otplib/blob/master/packages/otplib-core-async/README.md

@mattcollier  @dlongley

aljones15 commented on test/mocha/10-api.js:
```diff
@@ -1,3 +1,745 @@
+        await brAuthnToken.setAuthenticationRequirements({
+          account: accountId,
+          actor,
+          requiredAuthenticationMethods: ['token-client-registration'],
```
I believe in order to be multi-factor you need 2 authentication methods such as `password` and `otop` or `email-link` and `totp` etc. @dlongley am I wrong here?

Multifactor authentication:

https://www.nist.gov/itl/applied-cybersecurity/tig/back-basics-multi-factor-authentication

https://en.wikipedia.org/wiki/Multi-factor_authentication

@JSAssassin so I think you need 2 required authentication methods here, not one.

I might be wrong here as `token-client-registration` could be two auth methods in one.


EDIT: just for context the title of this test is:

```js
    it('should successfully login with multifactor authentication',
```


Is the second piece of authentication the new cookie that is set?

dlongley replied:
The API accepts any number of arguments. While best security practices are to have more than one, it would be awkward to force that here and it wouldn't cover all the use cases (or would require a separate API where only one is used).

aljones15 replied:
yes, but the test is called `should successfully login with multifactor authentication`, but it appears to only use one form of authentication. Is having one one `requiredAuthenticationMethods` and one authentication somehow multifactor here?

aljones15 commented on test/test.js:
```diff
@@ -2,10 +2,23 @@
 const bedrock = require('bedrock');
+const sinon = require('sinon');
+const brPassport = require('bedrock-passport');
+/* passportStub is defined globally here because we need to stub
```
If you want to try stubbing `bedrock-passport`'s `authenticate` method that might work, but it sounds like you have investigated this and know what you are doing so I 100% support this method especially if it saves on time.

aljones15 commented on test/package.json:
```diff
@@ -9,19 +9,26 @@
     "bedrock": "^3.1.1",
-    "bedrock-account": "^4.0.0",
-    "bedrock-authn-token": "^2.2.0",
+    "bedrock-account": "github:digitalbazaar/bedrock-account#unique-email-field",
```
has a release been made of `bedrock-account` with this feature?

JSAssassin replied:
nope, a release has not been made yet. I'm testing the current work on top level modules and making sure all these changes are tested where these modules are being used.

mattcollier replied:
You'll, probably circling back around to this, but noting this need updating.

JSAssassin replied:
Updated here cc8f8e0

aljones15 commented on test/package.json:
```diff
@@ -9,19 +9,26 @@
-    "bedrock-authn-token": "^2.2.0",
+    "bedrock-account": "github:digitalbazaar/bedrock-account#unique-email-field",
+    "bedrock-account-http": "^1.7.0",
+    "bedrock-authn-token": "github:digitalbazaar/bedrock-authn-token#delete-existing-nonce",
```
same issue here, has a release been made of this library?

JSAssassin committed:
Use `fix-verify` branch and add test to authenticate with password.

mattcollier commented on test/mocha/10-api.js:
```diff
@@ -425,6 +425,53 @@ describe('api', () => {
+      const password = 'test-password';
+      const saltRounds = 10;
+      const hash = await bcrypt.hash(password, saltRounds);
+      const {agent} = brHttpsAgent;
```
Just a note that with bedrock-https-agent@2, this can now be moved to the header along with other imports/requires.

JSAssassin replied:
Updated here 5ee7936

JSAssassin committed:
Update bedrock-account@5 and account-http@2 in test deps.

JSAssassin committed:
Use bedrock-https-agent@2.0 in test.

mattcollier reviewed (changes requested):
Please add a changelog entry.

JSAssassin committed:
Update deps and Add CHANGELOG.

mattcollier commented on CHANGELOG.md:
```diff
@@ -1,5 +1,12 @@
+## 2.0.0 - 2021-01-TBD
+
+### Changed
+- **BREAKING**: Update peerDependencies to include bedrock-account@5 and bedrock@4.
```
Can more be said about how/why upgrading the bedrock-account@5 constitutes a breaking change here?  I just did a quick pass and I don't see any new/different bedrock-account API calls.  And likewise, given that this module was not using any exotic logging features, the bedrock@4 bump does not make for a breaking change.

Folks internally and externally reading changelogs need to be able to assess what is going to be involved in upgrading from v1 to v2 here. So, we ask ourselves, if one wrote an application using v1, what is going to break if I simply change the dependency to v2?

A major bump in a peer dependency does not *necessarily* mean a breaking change here *unless* the breaking change in the peer dependency is somehow manifest in a public API change here.

JSAssassin replied:
okay, I think I understand, changing this to a minor release. 1f7e285

JSAssassin committed:
Update semver from major to minor release.

mattcollier commented on CHANGELOG.md:
```diff
@@ -1,5 +1,12 @@
+
+### Changed
+- Update peerDependencies to include bedrock-account@5 and bedrock@4.
+- Bake clientID into a cookie.
```
I believe the cookie handling is a new feature and therefore should be listed under an `### Added` heading.  A minor (feature) release will almost always include an added heading.

JSAssassin replied:
Updated here c305a77.

JSAssassin committed:
Add `Added` heading in CHANGELOG.
